### ISSUE TITLE
Figuring out front-end implementation of reply drafts

### ISSUE THREAD

**Issue by U-DON** [department-of-veterans-affairs/vets-api#375]
Some things I encountered while trying out replies and reply drafts…
1. When I start a reply to something in my Inbox, the thread has messages up to the last time the clinician responded. To use an example, message 621407 has two messages in its thread and the message itself is from the clinician. This is expected. When I save the draft of my reply to this message, the thread for the message draft now includes all of the user’s replies to this one message (621407). Sort of unexpected, but still sort of makes sense to me; if the user forgot to mention something in a previous reply, later replies will just get appended to that thread. Not saying this is a bug or anything, but wanted to see if this behavior is something we agree on.
   
   **TL;DR: Draft of reply to a message contains your previous replies to that message.**
2. When I send a reply from a draft, I ran into the same issues from before with duplicate sent messages. As a reminder, that issue occurred when sending replies to a thread where the last message was sent by yourself. And this results from the first point, where the draft’s thread has your most recent message as the last one in the thread.
   
   **TL;DR: Duplicate sent messages when replying to your own message are still happening.**
3. For replies, I’ve been using the last message in the thread for the id in the reply endpoint (`POST /messaging/health/messages/:id/reply`). Because of my first point, the last message in the thread for the draft is another message the user previously sent. So it ends up being used for the id for the reply; I don’t know if it’s correct to be using that id, or if the code on the front-end should search through the thread to find the last sender that wasn’t the user to determine which id to reply to.
   
   **TL;DR: Considering the first two points above, does front-end need to determine the id to reply to by a method other than just using the last message in the thread? Or is the behavior from the first point an issue?**

**Comment by mphprogrammer** [comment]
So, if I understand: Suppose I have a message in the inbox with id: 621407. I could send two replies, one after the other with ids 623563, followed by 628238. If I now create a new reply draft to 621407, with a new reply draft id of 657757. When I look at the thread for 657757, i would see the following, in order: [628238, 623563, 621407, ... ]. I would say the surest thing to do would be to search the thread for the first message on the list that has you as the recipient.

**Comment by mphprogrammer** [comment]
However, If I were to use another reply from the thread it still seems to work.

**Comment by U-DON** [comment]
Yup, the scenario is as you described.

I'm not sure where you were testing this out, though. Reply drafts implementation on the front-end isn't on staging yet; my observations were from my own branch that has yet to be merged.

**Comment by U-DON** [comment]
Another error I just came across when I tried saving a reply draft to message 573314 ("Release 16.2- SM last login" from Apr 13):

```
{"errors":[{"title":"Operation failed","detail":"Unable to reply because the source message is expired","code":"900","status":"400"}]}
```

**Comment by U-DON** [comment]
Filtering on `recipientName`/`recipientId` makes sense.

I was looking at the thread that seems to be growing way too big and noticed that the message you sent as the clinician wasn't showing up in the thread anymore, and then I realized that the default `per_page` for a thread is just 10.

Can this be increased to a much higher number? We would still have to figure out design-wise how we want to show threads that exceed the per-page number, but for now, increasing that limit should be enough.

**Comment by mphprogrammer** [comment]
We can increase the default per page - what number would be reasonable for the client?

**Comment by U-DON** [comment]
Perhaps 25? 50? Is there any way to always include the last message where the user was the recipient?

**Comment by ayaleloehr** [comment]
Let's do 50.  That seems conservative enough for now without having to change the logic to check for the last message the user received.

**Comment by U-DON** [comment]
50 seems conservative enough to me too. Similar to the edge case where a user wouldn't be able to see more than the first page of folders (I think that number was 100?), it seems highly unlikely to occur that they'd exceed the per-page number for messages, but we should probably think about handling these edge cases at some point.

**Comment by mphprogrammer** [comment]
I can set it 50, that's a small change - but also know that you are free to specify any page size using the per_page variable in your query.

**Comment by U-DON** [comment]
Actually, coming back to this, I'm not sure if replying to your own message is a problem as long as the clinician still sees it in the thread, right? Isn't the main issue that replying to your own message duplicating the sent message?

To re-use the example from above, [628238, 623563, 621407, ... ], where 621407 is the message you saw in your inbox and the first two ids are your replies; if you reply to your own message, 628238, would the clinician still see that reply along with 628238 and 623563 when they view the thread?

**Comment by U-DON** [comment]
To use an actual example for the behavior I'm observing (where your consecutive replies to an inbox message will all get added to the thread):

* I sent a reply to your 'some really new subject' message.
* When I view that reply in my Sent folder, I still see all of the test user's previous replies even though I specified the original message you sent as the message I was replying to.
* So it seems like I can reply to any message that came after that original message in the thread, and that reply would get appended to the history; the next time I reply to that original message, the previous reply will be included in the thread for the new reply.
* Even when I reply to my own message, the attributes for that reply has the clinician (not the test user) as the recipient.

To use much simpler id numbers for an example, let's say I got a message from a clinician with id 1.
* I reply to 1. Message 2, the reply, gets created.
* I reply to 1 again. Message 3 has in its thread [2, 1].
* I could reply to any of 1, 2, or 3, and the new message, 4, will have [3, 2, 1] as its thread.
* All 2, 3, and 4 correctly (?) have the clinician as the recipient no matter which message among [4, 3, 2, 1] I specified as the reply id.

**Comment by mphprogrammer** [comment]
I think you're right about that. As a side note, the same does not happen
for draft replies and threads. So If I create two consecutive draft replies
(ids 2 and 3) for message 1. Their respective history will only include
message 1.

**Comment by U-DON** [comment]
I can confirm seeing that for reply drafts. That behavior makes sense to me; they're not messages that haven't been officially sent yet, so only message 1 will show up in their histories. Only when they are sent as replies will they be included in each other's history.

It also appears that if you have 2 and 3 as drafts, sending 2 won't include 2 in 3's history. But then after sending 3, 2 will be in 3's history. Also makes sense.

**Comment by U-DON** [comment]
Closing this since the resolution is to discuss the strange behavior with MHV. The particular topic concerns duplication of sent messages when replying to the user's own message in a thread, even though the reply gets appended to the thread anyway just like a direct reply to the clinician's message.
